Src/EmbeddedScripts.CSharp.MonoEvaluator/MonoCSharpRunner.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.IO;
+using System.Threading.Tasks;
+using EmbeddedScripts.CSharp.MonoEvaluator.Hosting;
+using EmbeddedScripts.CSharp.Shared.CodeGeneration;
+using EmbeddedScripts.Shared;
+using EmbeddedScripts.Shared.Exceptions;
+using Mono.CSharp;
+
+namespace EmbeddedScripts.CSharp.MonoEvaluator
+{
+    public class MonoCSharpRunner : ICodeRunner, IEvaluator
+    {
+        private StringWriter _writer = new StringWriter();
+        private readonly Evaluator _evaluator;
+        private readonly int _runnerId = new Random().Next(0, int.MaxValue);
+
+        private void ThrowIfErrorIsLogged()
+        {
+            var message = _writer.ToString();
+
+            if (string.IsNullOrEmpty(message))
+                return;
+            
+            _writer.Close();
+            _writer = new StringWriter();
+                
+            throw new ScriptSyntaxErrorException(message);
+        }
+        
+        public MonoCSharpRunner()
+        {
+            var defaultSettings = new CompilerSettings();
+            _evaluator = new Evaluator(new CompilerContext(defaultSettings, new StreamReportPrinter(_writer)));
+            
+            HostFields.CreateNewEntry(_runnerId);
+
+            _evaluator.ReferenceAssembly(typeof(HostFields).Assembly);
+            
+            RunAsync("using EmbeddedScripts.CSharp.MonoEvaluator.Hosting;");
+        }
+
+        public Task RunAsync(string code)
+        {
+            _evaluator.Run(code);
+            
+            ThrowIfErrorIsLogged();
+
+            return Task.CompletedTask;
+        }
+
+        public ICodeRunner Register<T>(T obj, string alias)
+        {
+            HostFields.Get(_runnerId).Add(alias, obj);
+
+            var reflectionImporter = _evaluator.GetReflectionImporter();
+            var assembly = typeof(T).Assembly;
+
+            if (!reflectionImporter.ContainsAssembly(assembly))
```

Comment:
Is it for performance reasons? Will the code fail if you add the same assembly multiple times?

Reply:
No. If I reference assembly 2 or more times, Evaluator prints error `(1,25): error CS0433: The imported type `HelperObjects.HelperObject' is defined multiple times`. You can reproduce this just copy pasted line `_evaluator.ReferenceAssembly(assembly);` in `Register` method.